test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test.cc

```diff
@@ -996,8 +997,64 @@ TEST_P(GrpcJsonTranscoderIntegrationTest, EnableStrictRequestValidation) {
                                      {":path", "/unknown/path"},
                                      {":authority", "host"},
                                      {"content-type", "application/json"}},
-      "", {}, {}, Status(), Http::TestResponseHeaderMapImpl{{":status", "400"}},
-      "Bad request: Could not resolve /unknown/path to a method.", true, false, "", false);
+      "", {}, {}, Status(), Http::TestResponseHeaderMapImpl{{":status", "404"}},
+      "Could not resolve /unknown/path to a method.", true, false, "", false);
+
+  // Transcoding does not occur when unknown query param is included.
+  // HTTP Request to is passed directly to gRPC backend.
+  // gRPC response is passed directly to HTTP client.
+  testTranscoding<bookstore::GetShelfRequest, bookstore::Shelf>(
+      Http::TestRequestHeaderMapImpl{{":method", "GET"},
+                                     {":path", "/shelves/100?unknown=1"},
+                                     {":authority", "host"},
+                                     {"content-type", "application/json"}},
+      R"({ "theme" : "Children")", {}, {}, Status(Code::NOT_FOUND, "Shelf 9999 Not Found"),
+      Http::TestResponseHeaderMapImpl{
+          {":status", "200"}, {"grpc-status", "5"}, {"grpc-message", "Shelf 9999 Not Found"}},
+      "", true, false, R"({ "theme" : "Children")");
+}
+
+TEST_P(GrpcJsonTranscoderIntegrationTest, RejectUnknownQueryParam) {
+  const std::string filter =
+      R"EOF(
+            name: grpc_json_transcoder
+            typed_config:
+              "@type": type.googleapis.com/envoy.extensions.filters.http.grpc_json_transcoder.v3.GrpcJsonTranscoder
+              proto_descriptor : "{}"
+              services : "bookstore.Bookstore"
+              strict_http_request_validation:
+                reject_unknown_query_parameters: true
+            )EOF";
+  config_helper_.addFilter(
+      fmt::format(filter, TestEnvironment::runfilesPath("test/proto/bookstore.descriptor")));
+  HttpIntegrationTest::initialize();
+
+  // Transcoding does not occur from a request with the gRPC content type.
+  // We verify the request is not transcoded because the upstream receives the same JSON body.
+  // We verify the response is not transcoded because the HTTP status code does not match the gRPC
+  // status.
+  testTranscoding<bookstore::GetShelfRequest, bookstore::Shelf>(
+      Http::TestRequestHeaderMapImpl{{":method", "GET"},
+                                     {":path", "/shelves/100"},
+                                     {":authority", "host"},
+                                     {"content-type", "application/grpc"}},
```

Comment:
can we add grpc content-type case to all rejected cases?  to verify it is always pass-through due to content-type

Reply:
I actually have a question on this explicitly pass grpc content-type case.  I saw that this is added in the previous commit and it changes existing behavior too.  However, I am not sure if we really want to do it.

The argument is this:  This is a grpc_json transcoder or filter.  Therefore, if someone uses this filter, they should NOT pass in a request with grpc content-type.  If they indeed pass in grpc content-type, I think we should do something undefined or just continue to transcode it and if an error happens, let the error happens.  Because to me, this is a misconfig from the user side.

One interesting reason for this is, the grpc content-type cannot be really trusted.  Caller can set whatever content type they want, and this can allow them to use this to pass through a not real grpc request.

This basically adds another condition to the forwarding/transcoding rule, "if the content type is already grpc, then it will be passed through".  Do we really need this special handling?

Reply:
Yes,  always pass-through grpc content-type is a good feature to have so that this transcoder filter can be blindly added to the filter chain and the filter chain can support both grpc pass-through and json_transcoding.

Reply:
> Yes, always pass-through grpc content-type is a good feature to have so that this transcoder filter can be blindly added to the filter chain and the filter chain can support both grpc pass-through and json_transcoding.

This is exactly my concern.  Allow users to blindly add this filter to the filter chain and makes things seem working correctly is dangerous.

In what use case a user would send a request with grpc content-type and use this content type to let the filter pass through?  This seems to me a misconfigured envoy, especially when we can already control which paths/services be affected by the filter.

Reply:
> I actually have a question on this explicitly pass grpc content-type case. I saw that this is added in the previous commit and it changes existing behavior too. However, I am not sure if we really want to do it.

This was pre-existing behavior from before my PR #14715. My commit just moved the check outside a function call and added missing test cases for it (the filter has [low test coverage](https://github.com/envoyproxy/envoy/blob/ceddde2d682ffcb707ed1d36f9572e4f9c5feb39/test/per_file_coverage.sh#L37) currently). We do not want to remove this behavior, it would be a breaking change. github.com/GoogleCloudPlatform/esp-v2 currently depends on the behavior @qiwzhang mentioned.

> One interesting reason for this is, the grpc content-type cannot be really trusted. Caller can set whatever content type they want, and this can allow them to use this to pass through a not real grpc request.

In this case, we can add another option to reject requests with gRPC content type instead of passing them through. The design for the `RequestValidationOptions` field allows this extensibility now. This can be done in a follow-up PR.

> If they indeed pass in grpc content-type, I think we should do something undefined or just continue to transcode it and if an error happens, let the error happens. Because to me, this is a misconfig from the user side.

Any reason you prefer the undefined behavior over a strict rejection? We don't want users depending on undefined behavior we don't guarantee. https://www.hyrumslaw.com/

Reply:
You are right.  This was existing behavior.  I mis-read the change.  This is good then because anyway if we change it, it will be a breaking change.

I noticed this because when I debug, I can send arbitrary body with a fake grpc content type and force the filter to pass it through to backend.  This is kinda surprising to me.  But it is what it is now.

Thanks for the change!